# IBM Applied Data Science Capstone Course by Coursera
### Week 5 Final Report
**_Opening a New Gym Center in Singapore**
- Build a dataframe of neighborhoods in Singapore by web scraping the data from Wikipedia page
- Get the geographical coordinates of the neighborhoods
- Obtain the venue data for the neighborhoods from Foursquare API
- Explore and cluster the neighborhoods
- Select the best cluster to open a new Gym Center
***
### 1. Import libraries

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
!pip install geocoder
import geocoder # to get coordinates

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium
import folium # map rendering library

print("Libraries imported.")

     |████████████████████████████████| 98 kB 8.8 MB/s  eta 0:00:01
     |████████████████████████████████| 94 kB 3.5 MB/s eta 0:00:011
Libraries imported.


### 2. Scrap data from Wikipedia page into a DataFrame

In [3]:
# send the GET request
data = requests.get("https://en.wikipedia.org/wiki/Category:Places_in_Singapore").text

In [4]:
# parse data from the html into a beautifulsoup object
soup = BeautifulSoup(data, 'html.parser')

In [5]:
# create a list to store neighborhood data
neighborhoodList = []

In [6]:
# append the data into the list
for row in soup.find_all("div", class_="mw-category")[1].findAll("li"):
    neighborhoodList.append(row.text)

In [7]:
# create a new DataFrame from the list
sg_df = pd.DataFrame({"Neighborhood": neighborhoodList})

sg_df.head()

Neighborhood
0  List of commercial sites in Singapore
1            List of places in Singapore
2                   Admiralty, Singapore
3                 Airport Logistics Park
4                         Alkaff Gardens

In [8]:
# print the number of rows of the dataframe
sg_df.shape

(191, 1)

### 3. Get the geographical coordinates

In [9]:
# define a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Singapore'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [10]:
# call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(neighborhood) for neighborhood in sg_df["Neighborhood"].tolist() ]

In [11]:
coords

[[1.30215000000004, 103.86365000000006],
 [1.2904100000000653, 103.85211000000004],
 [1.4573513836356915, 103.81818372540098],
 [1.378020000000049, 104.00095000000005],
 [1.295220000000029, 103.85877000000005],
 [1.3716100000000324, 103.84546000000006],
 [1.3850100000000225, 103.84510000000006],
 [1.3023270067125625, 103.85813792658826],
 [1.3242500000000632, 103.95297000000005],
 [1.3327000000000453, 103.93714000000006],
 [1.3204607814059668, 103.86513527905811],
 [1.3195806402641834, 103.68613925693634],
 [1.295220000000029, 103.85877000000005],
 [1.3118300000000431, 103.79545000000007],
 [1.3507900000000745, 103.85110000000009],
 [1.31628967233386, 103.86329353979332],
 [1.3333300000000463, 103.70000000000005],
 [1.2904100000000653, 103.85211000000004],
 [1.38005674846275, 103.8918232554782],
 [1.3008100000000695, 103.85591000000005],
 [1.349520000000041, 103.75277000000006],
 [1.3482673252931419, 103.75171721711418],
 [1.2904100000000653, 103.85211000000004],
 [1.3641700000000583, 

In [12]:
# create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [13]:
# merge the coordinates into the original dataframe
sg_df['Latitude'] = df_coords['Latitude']
sg_df['Longitude'] = df_coords['Longitude']

In [14]:
# check the neighborhoods and the coordinates
print(sg_df.shape)
sg_df

(191, 3)


Neighborhood  Latitude   Longitude
0    List of commercial sites in Singapore  1.302150  103.863650
1              List of places in Singapore  1.290410  103.852110
2                     Admiralty, Singapore  1.457351  103.818184
3                   Airport Logistics Park  1.378020  104.000950
4                           Alkaff Gardens  1.295220  103.858770
5                               Ang Mo Kio  1.371610  103.845460
6               Ang Mo Kio Police Division  1.385010  103.845100
7                              Arab Street  1.302327  103.858138
8                                    Bedok  1.324250  103.952970
9                    Bedok Police Division  1.332700  103.937140
10                    Bendemeer, Singapore  1.320461  103.865135
11                            Benoi Sector  1.319581  103.686139
12                         Bidadari Garden  1.295220  103.858770
13                           Bidadari Park  1.311830  103.795450
14                       Bishan, Singapore  1.350790  103.851100
15                               Boon Keng  1.316290  103.863294
16                                Boon Lay  1.333330  103.700000
17                   Brickworks, Singapore  1.290410  103.852110
18                                Buangkok  1.380057  103.891823
19                        Bugis, Singapore  1.300810  103.855910
20                             Bukit Batok  1.349520  103.752770
21                     Bukit Batok Central  1.348267  103.751717
22                            Bukit Chandu  1.290410  103.852110
23                            Bukit Gombak  1.364170  103.757220
24                             Bukit Merah  1.284170  103.823060
25                           Bukit Panjang  1.378770  103.769770
26                            Bukit Purmei  1.274016  103.828384
27                             Bukit Timah  1.340410  103.772210
28                        Bukit Timah Hill  1.354700  103.776400
29                             Buona Vista  1.280560  103.788060
30                          Caldecott Hill  1.335400  103.836610
31               Canberra Plaza, Singapore  1.442930  103.830190
32                            Cashin House  1.296416  103.853475
33                 Central Area, Singapore  1.288830  103.846250
34                 Central Police Division  1.278930  103.839090
35                                  Changi  1.355140  103.990060
36                          Changi Village  1.384440  103.983060
37                                Chin Bee  1.319330  103.662100
38                           Choa Chu Kang  1.386160  103.746180
39                    City Hall, Singapore  1.292570  103.856870
40                Clementi Police Division  1.317170  103.766560
41                     Clementi, Singapore  1.314380  103.765370
42                Coleman House, Singapore  1.292080  103.850393
43                 Commonwealth, Singapore  1.304798  103.797610
44                     Crawford, Singapore  1.334710  103.903230
45                    Defu Industrial Park  1.360932  103.897111
46                             Dhoby Ghaut  1.298740  103.848000
47                        Dover, Singapore  1.314370  103.893460
48                 Emerald Hill, Singapore  1.303708  103.839567
49                    Esplanade, Singapore  1.288890  103.854261
50                             Farrer Park  1.312220  103.848890
51                       Farrer Park Field  1.311440  103.849120
52                           Gek Poh Ville  1.347013  103.697949
53                                 Geylang  1.311470  103.882180
54                           Geylang Bahru  1.322191  103.870579
55                            Geylang East  1.319790  103.886770
56                                Ghim Moh  1.311110  103.787780
57                       Guilin, Singapore  1.290410  103.852110
58                              Gul Circle  1.314699  103.663578
59                HarbourFront (Singapore)  1.265331  103.818861
60                     Hillview, Singapore  1.362161  103.765872


In [15]:
# save the DataFrame as CSV file
sg_df.to_csv("sg_df.csv", index=False)

### 4. Create a map of Singapore with neighborhoods superimposed on top

In [16]:
# get the coordinates of Singapore
address = 'Singapore'

geolocator = Nominatim(user_agent="ny-explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Singapore {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Singapore 1.357107, 103.8194992.


In [17]:
# create map of Singapore using latitude and longitude values
map_sg = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(sg_df['Latitude'], sg_df['Longitude'], sg_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_sg)  
    
map_sg

In [18]:
# save the map as HTML file
map_sg.save('map_sg.html')

### 5. Use the Foursquare API to explore the neighborhoods

In [19]:
# define Foursquare Credentials and Version
CLIENT_ID = 'RY5RKWDUYRZS50JJI3XX3BJ3PJPKZUMPIC34NQF42Q1F1VUC' # your Foursquare ID
CLIENT_SECRET = 'XGF4K35X4QQCXMZNHDC100O1CVBTGMOGOC1XJXARVQ3NVF3L' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: RY5RKWDUYRZS50JJI3XX3BJ3PJPKZUMPIC34NQF42Q1F1VUC
CLIENT_SECRET:XGF4K35X4QQCXMZNHDC100O1CVBTGMOGOC1XJXARVQ3NVF3L


**Now, let's get the top 100 venues that are within a radius of 1000 meters.**

In [116]:
radius = 1000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(sg_df['Latitude'], sg_df['Longitude'], sg_df['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [117]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(12306, 7)


Neighborhood  Latitude  Longitude  \
0  List of commercial sites in Singapore   1.30215  103.86365   
1  List of commercial sites in Singapore   1.30215  103.86365   
2  List of commercial sites in Singapore   1.30215  103.86365   
3  List of commercial sites in Singapore   1.30215  103.86365   
4  List of commercial sites in Singapore   1.30215  103.86365   

              VenueName  VenueLatitude  VenueLongitude          VenueCategory  
0         The Projector       1.302255      103.863553              Multiplex  
1           Wakey Wakey       1.301162      103.861926            Coffee Shop  
2  Sari Ratu Restaurant       1.302145      103.861258  Indonesian Restaurant  
3               Mrs Pho       1.301323      103.861880  Vietnamese Restaurant  
4             PizzaFace       1.301028      103.861535     Italian Restaurant

**Let's check how many venues were returned for each neighorhood**

In [118]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  \
Neighborhood                                                            
Admiralty, Singapore                         21         21         21   
Airport Logistics Park                       15         15         15   
Alkaff Gardens                              100        100        100   
Ang Mo Kio                                   80         80         80   
Ang Mo Kio Police Division                   27         27         27   
Arab Street                                 100        100        100   
Bedok                                        34         34         34   
Bedok Police Division                        68         68         68   
Bendemeer, Singapore                         51         51         51   
Benoi Sector                                  6          6          6   
Bidadari Garden                             100        100        100   
Bidadari Park                                82         82         82   
Bishan, Singapore                            70         70         70   
Boon Keng                                    99         99         99   
Boon Lay                                     47         47         47   
Brickworks, Singapore                       100        100        100   
Buangkok                                     51         51         51   
Bugis, Singapore                            100        100        100   
Bukit Batok                                  49         49         49   
Bukit Batok Central                          46         46         46   
Bukit Chandu                                100        100        100   
Bukit Gombak                                 50         50         50   
Bukit Merah                                  93         93         93   
Bukit Panjang                                69         69         69   
Bukit Purmei                                 42         42         42   
Bukit Timah                                  89         89         89   
Bukit Timah Hill                             30         30         30   
Buona Vista                                  31         31         31   
Caldecott Hill                               37         37         37   
Canberra Plaza, Singapore                    36         36         36   
Cashin House                                100        100        100   
Central Area, Singapore                     100        100        100   
Central Police Division                     100        100        100   
Changi                                       60         60         60   
Changi Village                               60         60         60   
Chin Bee                                      8          8          8   
Choa Chu Kang                                36         36         36   
City Hall, Singapore                        100        100        100   
Clementi Police Division                     90         90         90   
Clementi, Singapore                          83         83         83   
Coleman House, Singapore                    100        100        100   
Commonwealth, Singapore                      95         95         95   
Crawford, Singapore                          18         18         18   
Defu Industrial Park                         23         23         23   
Dhoby Ghaut                                 100        100        100   
Dover, Singapore                            100        100        100   
Emerald Hill, Singapore                     100        100        100   
Esplanade, Singapore                        100        100        100   
Farrer Park                                 100        100        100   
Farrer Park Field                           100        100        100   
Gek Poh Ville                                25         25         25   
Geylang                                     100        100        100   
Geylang Bahru                                58         58         58   
Geylang East            

**Let's find out how many unique categories can be curated from all the returned venues**

In [119]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 355 uniques categories.


In [120]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]

array(['Multiplex', 'Coffee Shop', 'Indonesian Restaurant',
       'Vietnamese Restaurant', 'Italian Restaurant', 'Thai Restaurant',
       'Burger Joint', 'Cafeteria', 'Hotel', 'BBQ Joint', 'Art Gallery',
       'Café', 'Food Court', 'Restaurant', 'French Restaurant', 'Bakery',
       'Yoga Studio', 'Sandwich Place', 'Tea Room',
       'Taiwanese Restaurant', 'Cocktail Bar', 'Indian Restaurant',
       'Pool', 'Hookah Bar', 'Bed & Breakfast', 'Asian Restaurant',
       'Whisky Bar', 'Spa', 'Park', 'Beer Bar', 'Japanese Restaurant',
       'Bar', 'Jazz Club', 'Vegetarian / Vegan Restaurant',
       'New American Restaurant', 'Wine Bar', 'Mediterranean Restaurant',
       'Mexican Restaurant', 'Movie Theater', 'Racetrack', 'Supermarket',
       'Halal Restaurant', 'Bridge', 'Dessert Shop', 'Chinese Restaurant',
       'Shopping Mall', 'Gym / Fitness Center', 'Ramen Restaurant',
       'Dim Sum Restaurant', 'Arts & Crafts Store'], dtype=object)

In [121]:
# check if the results contain "Gym/Fitness Center"
"Gym / Fitness Center" in venues_df['VenueCategory'].unique()

True

### 6. Analyze Each Neighborhood

In [122]:
# one hot encoding
sg_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
sg_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [sg_onehot.columns[-1]] + list(sg_onehot.columns[:-1])
sg_onehot = sg_onehot[fixed_columns]

print(sg_onehot.shape)
sg_onehot.head()

(12306, 356)


Neighborhoods  ATM  Accessories Store  Airport  \
0  List of commercial sites in Singapore    0                  0        0   
1  List of commercial sites in Singapore    0                  0        0   
2  List of commercial sites in Singapore    0                  0        0   
3  List of commercial sites in Singapore    0                  0        0   
4  List of commercial sites in Singapore    0                  0        0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Aquarium  Arcade  Art Gallery  Art Museum  \
0                    0         0       0            0           0   
1                    0         0       0            0           0   
2                    0         0       0            0           0   
3                    0         0       0            0           0   
4                    0         0       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0                    0                 0                   0   
1                    0                 0                   0   
2                    0                 0                   0   
3                    0                 0                   0   
4                    0                 0                   0   

   Australian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                      0            0              0                0   
1                      0            0              0                0   
2                      0            0              0                0   
3                      0            0              0                0   
4                      0            0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Basketball Court  Bay  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
0                 0    0      0          0                0         0   
1                 0    0      0          0                0         0   
2                 0    0      0          0                0         0   
3                 0    0      0          0                0         0   
4                 0    0      0          0                0         0   

   Beer Garden  Beer Store  Betting Shop  Big Box Store  \
0            0           0             0              0   
1            0           0             0              0   
2            0           0             0              0   
3            0           0             0              0   
4            0           0             0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boarding House  Boat or Ferry  Bookstore  Border Crossing  \
0               0              0          0                0   
1               0              0          0                0   
2               0     

**Count the number of Gym/Fitness centers for each neighborhood**

In [124]:
gym_venues_df = sg_onehot.loc[sg_onehot["Gym / Fitness Center"]>0, ["Neighborhoods", "Gym / Fitness Center"]]
gym_count = gym_venues_df.groupby(["Neighborhoods"])["Gym / Fitness Center"].count().reset_index()
gym_count.rename(columns={"Neighborhoods": "Neighborhood", "Gym / Fitness Center": "No. of Gym/Fitness Center"}, inplace=True)
print(gym_count.shape)
gym_count.head()

(68, 2)


Neighborhood  No. of Gym/Fitness Center
0              Alkaff Gardens                          2
1                  Ang Mo Kio                          1
2  Ang Mo Kio Police Division                          1
3                 Arab Street                          1
4             Bidadari Garden                          2

**Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category**

In [125]:
sg_grouped = sg_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(sg_grouped.shape)
sg_grouped

(191, 356)


Neighborhoods   ATM  Accessories Store   Airport  \
0                     Admiralty, Singapore  0.00           0.000000  0.000000   
1                   Airport Logistics Park  0.00           0.000000  0.133333   
2                           Alkaff Gardens  0.00           0.000000  0.000000   
3                               Ang Mo Kio  0.00           0.000000  0.000000   
4               Ang Mo Kio Police Division  0.00           0.000000  0.000000   
5                              Arab Street  0.00           0.000000  0.000000   
6                                    Bedok  0.00           0.000000  0.000000   
7                    Bedok Police Division  0.00           0.000000  0.000000   
8                     Bendemeer, Singapore  0.00           0.000000  0.000000   
9                             Benoi Sector  0.00           0.000000  0.000000   
10                         Bidadari Garden  0.00           0.000000  0.000000   
11                           Bidadari Park  0.00           0.000000  0.000000   
12                       Bishan, Singapore  0.00           0.000000  0.000000   
13                               Boon Keng  0.00           0.000000  0.000000   
14                                Boon Lay  0.00           0.000000  0.000000   
15                   Brickworks, Singapore  0.00           0.000000  0.000000   
16                                Buangkok  0.00           0.000000  0.000000   
17                        Bugis, Singapore  0.00           0.000000  0.000000   
18                             Bukit Batok  0.00           0.000000  0.000000   
19                     Bukit Batok Central  0.00           0.000000  0.000000   
20                            Bukit Chandu  0.00           0.000000  0.000000   
21                            Bukit Gombak  0.02           0.000000  0.000000   
22                             Bukit Merah  0.00           0.000000  0.000000   
23                           Bukit Panjang  0.00           0.000000  0.000000   
24                            Bukit Purmei  0.00           0.000000  0.000000   
25                             Bukit Timah  0.00           0.000000  0.000000   
26                        Bukit Timah Hill  0.00           0.000000  0.000000   
27                             Buona Vista  0.00           0.000000  0.000000   
28                          Caldecott Hill  0.00           0.000000  0.000000   
29               Canberra Plaza, Singapore  0.00           0.000000  0.000000   
30                            Cashin House  0.00           0.000000  0.000000   
31                 Central Area, Singapore  0.00           0.000000  0.000000   
32                 Central Police Division  0.00           0.000000  0.000000   
33                                  Changi  0.00           0.000000  0.066667   
34                          Changi Village  0.00           0.000000  0.000000   
35                                Chin Bee  0.00           0.000000  0.000000   
36                           Choa Chu Kang  0.00           0.000000  0.000000   
37                    City Hall, Singapore  0.00           0.000000  0.000000   
38                Clementi Police Division  0.00           0.000000  0.000000   
39                     Clementi, Singapore  0.00           0.000000  0.000000   
40                Coleman House, Singapore  0.00           0.000000  0.000000   
41                 Commonwealth, Singapore  0.00           0.000000  0.000000   
42                     Crawford, Singapore  0.00           0.000000  0.000000   
43                    Defu Industrial Park  0.00           0.000000  0.043478   
44                             Dhoby Ghaut  0.00           0.000000  0.000000   
45                        Dover, Singapore  0.00           0.000000  0.000000   
46                 Emerald Hill, Singapore  0.00           0.000000  0.000000   
47                    Esplanade, Singapore  0.00           0.000000  0.000000   
48                             Farrer Park  0.00           0

In [126]:
len(sg_grouped[sg_grouped["Gym / Fitness Center"] > 0])

68

**Create a new DataFrame for Gym/Fitness Center data only**

In [127]:
sg_gym = sg_grouped[["Neighborhoods","Gym / Fitness Center"]]

In [134]:
sg_gym.head()
sg_gym.shape

(191, 2)

### 7. Cluster Neighborhoods
Run k-means to cluster the neighborhoods in Singapore into 3 clusters.

In [136]:
# set number of clusters
kclusters = 3

sg_clustering = sg_gym.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(sg_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

191

In [151]:
# create a new dataframe that includes the cluster.
sg_merged = sg_gym.copy()

# add clustering labels
sg_merged["Cluster Labels"] = kmeans.labels_
sg_merged.shape

(191, 3)

In [152]:
sg_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
sg_merged.head()
sg_merged.shape

(191, 3)

In [153]:
# merge sg_merged with gym_count and sg_df to add number of gym centers and latitude/longitude for each neighborhood
sg_merged = pd.merge(sg_merged, gym_count, on="Neighborhood", how="left")
sg_merged["No. of Gym/Fitness Center"] = sg_merged["No. of Gym/Fitness Center"].replace(np.nan, 0)
sg_merged = sg_merged.join(sg_df.set_index("Neighborhood"), on="Neighborhood")

print(sg_merged.shape)
sg_merged.head()

(191, 6)


Neighborhood  Gym / Fitness Center  Cluster Labels  \
0        Admiralty, Singapore              0.000000               0   
1      Airport Logistics Park              0.000000               0   
2              Alkaff Gardens              0.020000               2   
3                  Ang Mo Kio              0.012500               2   
4  Ang Mo Kio Police Division              0.037037               1   

   No. of Gym/Fitness Center  Latitude   Longitude  
0                        0.0  1.457351  103.818184  
1                        0.0  1.378020  104.000950  
2                        2.0  1.295220  103.858770  
3                        1.0  1.371610  103.845460  
4                        1.0  1.385010  103.845100

In [154]:
# sort the results by Cluster Labels
print(sg_merged.shape)
sg_merged.sort_values(["Cluster Labels"], inplace=True)
sg_merged

(191, 6)


Neighborhood  Gym / Fitness Center  \
0                     Admiralty, Singapore              0.000000   
91                    Kembangan, Singapore              0.000000   
90                              Kebun Baru              0.000000   
89                                  Katong              0.000000   
88                             Kampong Ubi              0.000000   
144                              Sembawang              0.000000   
86                            Kampong Glam              0.000000   
85                           Kampong Bugis              0.000000   
145              Sembawang Shopping Centre              0.000000   
83                           Kallang Bahru              0.000000   
82                                 Kallang              0.000000   
81                   Kaki Bukit, Singapore              0.000000   
147                                 Senoko              0.000000   
148                              Serangoon              0.000000   
149                       Serangoon Garden              0.000000   
77                             Jurong Port              0.000000   
76                  Jurong Police Division              0.000000   
75                             Jurong Hill              0.000000   
150                        Serangoon North              0.000000   
73                                  Jurong              0.000000   
72                                Joo Seng              0.000000   
71                                Joo Koon              0.000000   
70                          Joo Chiat Road              0.000000   
153                                  Simei              0.000000   
68                              Jalan Kayu              0.000000   
67                             Jalan Besar              0.000000   
66                             Jalan Bahar              0.000000   
65                         Istana Woodneuk              0.000000   
143                                Seletar              0.000000   
141                   Rivervale, Singapore              0.000000   
94                           Kim Seng Road              0.000000   
140                River Valley, Singapore              0.000000   
123                           Orchard Road              0.000000   
126                         Pandan Gardens              0.000000   
121          Old Police Academy, Singapore              0.000000   
120                      Novena, Singapore              0.000000   
128                              Pasir Ris              0.000000   
129                             Paya Lebar              0.000000   
130                       Penjuru Crescent              0.000000   
116                Mount Vernon, Singapore              0.000000   
115                       Mount Emily Park              0.000000   
131                        Peranakan Place              0.000000   
133                         Pioneer Sector              0.000000   
112                              Marsiling              0.000000   
111                          Marine Parade              0.000000   
64                         Istana Tyersall              0.000000   
134                           Potong Pasir              0.000000   
108                  Marina Bay, Singapore              0.000000   
107                                 Mandai              0.000000   
106                  MacPherson, Singapore              0.000000   
105                      Loyang, Singapore              0.000000   
104                          Lower Seletar              0.000000   
103                           Lorong Halus              0.000000   
102                           Lorong Chuan              0.000000   
101                Little India, Singapore              0.000000   
136                     Punggol Point Park              0.000000   
138                              Radin Mas              0.000000   
98                            Lim Chu Kang              0.000000   
97                

**Finally, let's visualize the resulting clusters**

In [155]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(sg_merged['Latitude'], sg_merged['Longitude'], sg_merged['Neighborhood'], sg_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [156]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

### 8. Examine Clusters

#### Cluster 0

In [161]:
cluster_0 = sg_merged.loc[sg_merged['Cluster Labels'] == 0]
print(cluster_0.shape[0])
cluster_0

123


Neighborhood  Gym / Fitness Center  Cluster Labels  \
0             Admiralty, Singapore                   0.0               0   
91            Kembangan, Singapore                   0.0               0   
90                      Kebun Baru                   0.0               0   
89                          Katong                   0.0               0   
88                     Kampong Ubi                   0.0               0   
144                      Sembawang                   0.0               0   
86                    Kampong Glam                   0.0               0   
85                   Kampong Bugis                   0.0               0   
145      Sembawang Shopping Centre                   0.0               0   
83                   Kallang Bahru                   0.0               0   
82                         Kallang                   0.0               0   
81           Kaki Bukit, Singapore                   0.0               0   
147                         Senoko                   0.0               0   
148                      Serangoon                   0.0               0   
149               Serangoon Garden                   0.0               0   
77                     Jurong Port                   0.0               0   
76          Jurong Police Division                   0.0               0   
75                     Jurong Hill                   0.0               0   
150                Serangoon North                   0.0               0   
73                          Jurong                   0.0               0   
72                        Joo Seng                   0.0               0   
71                        Joo Koon                   0.0               0   
70                  Joo Chiat Road                   0.0               0   
153                          Simei                   0.0               0   
68                      Jalan Kayu                   0.0               0   
67                     Jalan Besar                   0.0               0   
66                     Jalan Bahar                   0.0               0   
65                 Istana Woodneuk                   0.0               0   
143                        Seletar                   0.0               0   
141           Rivervale, Singapore                   0.0               0   
94                   Kim Seng Road                   0.0               0   
140        River Valley, Singapore                   0.0               0   
123                   Orchard Road                   0.0               0   
126                 Pandan Gardens                   0.0               0   
121  Old Police Academy, Singapore                   0.0               0   
120              Novena, Singapore                   0.0               0   
128                      Pasir Ris                   0.0               0   
129                     Paya Lebar                   0.0               0   
130               Penjuru Crescent                   0.0               0   
116        Mount Vernon, Singapore                   0.0               0   
115               Mount Emily Park                   0.0               0   
131                Peranakan Place                   0.0               0   
133                 Pioneer Sector                   0.0               0   
112                      Marsiling                   0.0               0   
111                  Marine Parade                   0.0               0   
64                 Istana Tyersall                   0.0               0   
134                   Potong Pasir                   0.0               0   
108          Marina Bay, Singapore                   0.0               0   
107                         Mandai                   0.0               0   
106          MacPherson, Singapore                   0.0               0   
105              Loyang, Singapore                   0.0               0   
104                  Lower Seletar                   0.0              

#### Cluster 1

In [162]:
cluster_1 = sg_merged.loc[sg_merged['Cluster Labels'] == 1]
print(cluster_1.shape[0])
cluster_1

8


Neighborhood  Gym / Fitness Center  Cluster Labels  \
182  Western Islands, Singapore              0.062500               1   
166           Telok Ayer Street              0.050000               1   
186                     Yew Tee              0.071429               1   
190           Yunnan, Singapore              0.043478               1   
93                    Kian Teck              0.055556               1   
31      Central Area, Singapore              0.040000               1   
4    Ang Mo Kio Police Division              0.037037               1   
14                     Boon Lay              0.042553               1   

     No. of Gym/Fitness Center  Latitude   Longitude  
182                        1.0  1.331100  103.695510  
166                        5.0  1.280960  103.847752  
186                        2.0  1.394440  103.753890  
190                        1.0  1.337348  103.690630  
93                         1.0  1.332837  103.695545  
31                         4.0  1.288830  103.846250  
4                          1.0  1.385010  103.845100  
14                         2.0  1.333330  103.700000

#### Cluster 2

In [163]:
cluster_2 = sg_merged.loc[sg_merged['Cluster Labels'] == 2]
print(cluster_2.shape[0])
cluster_2

60


Neighborhood  Gym / Fitness Center  \
127                          Pasir Panjang              0.029412   
32                 Central Police Division              0.030000   
169                       Template:Punggol              0.016393   
168           Template:Places in Singapore              0.020000   
36                           Choa Chu Kang              0.027778   
37                    City Hall, Singapore              0.010000   
40                Coleman House, Singapore              0.010000   
164                            Tanjong Rhu              0.016129   
41                 Commonwealth, Singapore              0.010526   
44                             Dhoby Ghaut              0.010000   
47                    Esplanade, Singapore              0.010000   
30                            Cashin House              0.020000   
25                             Bukit Timah              0.011236   
23                           Bukit Panjang              0.014493   
51                                 Geylang              0.010000   
20                            Bukit Chandu              0.020000   
17                        Bugis, Singapore              0.010000   
15                   Brickworks, Singapore              0.020000   
179                 Waterboat House Garden              0.030000   
11                           Bidadari Park              0.012195   
10                         Bidadari Garden              0.020000   
5                              Arab Street              0.010000   
185               Yan Kit Swimming Complex              0.011364   
3                               Ang Mo Kio              0.012500   
2                           Alkaff Gardens              0.020000   
188                                 Yishun              0.012500   
174                               Toh Tuck              0.011111   
54                                Ghim Moh              0.010000   
55                       Guilin, Singapore              0.020000   
157                           Straits View              0.030000   
119                      Newton, Singapore              0.010000   
118                   Museum Planning Area              0.020000   
117                 Mountbatten, Singapore              0.010000   
114                     Millenia Singapore              0.010000   
132                           Pioneer Mall              0.018519   
113                           Merlion Park              0.030000   
110                           Marina South              0.010000   
135                                Punggol              0.016393   
100            List of places in Singapore              0.020000   
137                  Queenstown, Singapore              0.015873   
99   List of commercial sites in Singapore              0.010000   
139                          Raffles House              0.010000   
189                       Yuhua, Singapore              0.020000   
122                              One-north              0.025974   
142                                  Safti              0.020000   
87                            Kampong Java              0.010101   
84                            Kallang Park              0.011364   
146                               Sengkang              0.025641   
80                     Jurong West Central              0.012500   
79                             Jurong West              0.013699   
78                        Jurong Town Hall              0.010989   
74                             Jurong East              0.010870   
151                            Shenton Way              0.030000   
152                                 Siglap              0.011364   
69                    Jalan Kubor Cemetery              0.010000   
154                                Simpang              0.013889   
59              Holland Village, Singapore              0.010753   
156                 Singapore Science Park              0.013514   
92                

#### Observations:
There are 123 out of the 191 neighborhoods fall into cluster 0 which showing as red circle on the map. Cluster 0 provides a catergory of very low frequency of occurance of Gym and Fitness Center in the neighborhoods. In fact, there is zero existance of Gym and Fitness Centers in these areas. 60 neighborhoods in green in cluster 2 is considered as as moderate level of frequency of occurance. The balance of 8 neighborhoods in cluster 1 as purple shows the highest frequency of occurance of Gym and Fitness Center in the neighborhoods.